citation-server/src/graphql/query.js

```diff
@@ -20,71 +21,119 @@ export const ObjectInterface = new GraphQLInterfaceType({
 	})
 });
 
-export const PageType = new GraphQLObjectType({
-	name: 'Page',
-	interfaces: [ObjectInterface],
-	isTypeOf: value => value.__type__ === 'Page',
-	fields: () => ({
-		__id__: {type: GraphQLID},
-		__type__: {type: GraphQLString},
-		__tree__: {
-			type: GraphQLString,
-			resolve: inspect
-		},
-		__newId__: {type: GraphQLID},
-		slug: {type: GraphQLString},
-		title: {type: GraphQLString},
-		children: {
-			type: new GraphQLList(PageType),
-			resolve: root => readChildren(root.children)
-		},
-		component: {
-			type: ComponentType,
-			resolve: root => readChild(root.component)
-		}
-	})
-});
-
-export const ComponentType = new GraphQLObjectType({
-	name: 'Component',
-	interfaces: [ObjectInterface],
-	isTypeOf: value => value.__type__ === 'Component',
-	fields: () => ({
-		__id__: {type: GraphQLID},
-		__type__: {type: GraphQLString},
-		__tree__: {
-			type: GraphQLString,
-			resolve: inspect
-		},
-		__newId__: {type: GraphQLID},
-		type: {type: GraphQLString},
-		children: {
-			type: new GraphQLList(ComponentType),
-			resolve: root => readChildren(root.children)
-		},
-		data: {
-			type: new GraphQLList(ObjectInterface),
-			resolve: root => readChildren(root.data)
-		}
-	})
-});
+export async function buildObjects() {
+	const ObjectTypes = {};
+	const model = await readModel();
+	for (const structure of model) {
+		ObjectTypes[structure.name] = new GraphQLObjectType({
+			name: structure.name,
+			interfaces: [ObjectInterface],
+			isTypeOf: value => value.__type__ === structure.name,
+			fields: () => {
+				const resultFields = {
+					__id__: {type: GraphQLID},
+					__newId__: {type: GraphQLID},
+					__type__: {type: GraphQLString},
+					__tree__: {
+						type: GraphQLString,
+						resolve: inspect
+					}
+				};
+				structure.fields.forEach(field => {
+					switch (field.type) {
```

Comment:
This switch with a field which can be a string or an array seems dangerous to me. Ex if someone use a type not defined, the system automatically consider the value is an array and will fail.

I would prefer a first if on the type of field.type (string or array) and then, a switch for both case, one on field.type, another on field.type[0]. Each will have a default for when we didn't recognize the type asked.